File: 3867. sum-of-gcd-of-formed-pairs.java
import java.util.Arrays;

class Solution {
    public long gcdSum(int[] nums) {
        int n = nums.length;
        int mx = 0;

        for (int i = 0; i < n; i++) {
            mx = Math.max(mx, nums[i]);
            nums[i] = gcd(mx, nums[i]);
        }

        Arrays.sort(nums);

        int left = 0, right = n - 1;
        long ans = 0;

        while (left < right) {
            ans += gcd(nums[left], nums[right]);
            left++;
            right--;
        }

        return ans;
    }

    private int gcd(int a, int b) {
        while (b != 0) {
            int t = a % b;
            a = b;
            b = t;
        }
        return a;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();

        // Example 1
        int[] nums1 = {2, 6, 4};
        System.out.println(sol.gcdSum(nums1)); // 2

        // Example 2
        int[] nums2 = {3, 6, 2, 8};
        System.out.println(sol.gcdSum(nums2)); // 5
    }
}

File: 3867. sum-of-gcd-of-formed-pairs.py
from math import gcd


class Solution:
    def gcdSum(self, nums: list[int]) -> int:
        mx = 0

        for i in range(len(nums)):
            mx = max(mx, nums[i])
            nums[i] = gcd(mx, nums[i])

        nums.sort()

        left, right = 0, len(nums) - 1
        ans = 0

        while left < right:
            ans += gcd(nums[left], nums[right])
            left += 1
            right -= 1

        return ans


if __name__ == "__main__":
    sol = Solution()

    # Example 1
    nums = [2, 6, 4]
    print(sol.gcdSum(nums))  # 2

    # Example 2
    nums = [3, 6, 2, 8]
    print(sol.gcdSum(nums))  # 5